### PR TITLE
Made a cutomized version of sdist command to use cython

### PR DESCRIPTION
**Pull request by unnonouno**
We need to use Cython before making sdist. I customized `sdist` command to call `cythonize`. `Command.sub_commands` are called before its `run` method.

### CODE REVIEW

**Comment by sonots** [comment]
Ah, I was checking sdist after running install. This change is safer and easy to use. LGTM!

**Comment by sonots** [comment]
travis is failing as:

```
/tmp/tmpcDCLTu/a.cpp:1:23: fatal error: cublas_v2.h: No such file or directory
 #include <cublas_v2.h>
```

**Comment by sonots** [comment]
It seems the cause `--cupy-no-cuda` is not working is because I modified to call `get_ext_modules` twice (one is for `ext_modules` and one is in `build_ext`).

```
 95     ext_modules=ext_modules,
 96     cmdclass={'build_ext': build_ext}
```

and `parse_args` removes `sys.argv`. The direct easy way to fix is to make parsed args be global variable.

```diff
diff --git a/cupy_setup_build.py b/cupy_setup_build.py
index bb1a942..c99c24a 100644
--- a/cupy_setup_build.py
+++ b/cupy_setup_build.py
@@ -319,6 +319,10 @@ def parse_args():
     return arg_options


+cupy_setup_options = parse_args()
+print('Options:', cupy_setup_options)
+
+
 def check_cython_version():
     try:
         import Cython
@@ -356,8 +360,7 @@ def check_extensions(extensions):


 def get_ext_modules(use_cython=False):
-    arg_options = parse_args()
-    print('Options:', arg_options)
+    arg_options = cupy_setup_options

     # We need to call get_config_vars to initialize _config_vars in distutils
     # see #1849
```

**Comment by unnonouno** [comment]
I see. I feel it is enough.

File: cupy_setup_build.py [hunk 1]
```diff
@@ -318,6 +319,9 @@ def parse_args():
     return arg_options
 
 
+cupy_setup_otions = parse_args()
```
Comment (sonots) [inline cupy_setup_build.py]: typo

**Comment by sonots** [comment]
With global `cupy_setup_options`, we are possible to eliminate `arg_options` arguments from `cythonize` and `make_extensions`. What do you think?

**Comment by sonots** [comment]
I could reproduce SEGV on my no GPU linux machine.

I could avoid SEGV by running `python setup.py sdist --cupy-no-cuda` twice.
It seems `rm -r .eggs` reproduces SEGV again. So,

* install `fastrlock` then `python setup.py sdist --cupy-no-cuda` => No error
* install `fastrlock` during `python setup.py sdist --cupy-no-cuda` with `setup_requires` => SEGV

**Comment by sonots** [comment]
SEGV did not happen when I run `python setup.py sdist --cupy-no-cuda` on a GPU machine.
Hmm...

**Comment by sonots** [comment]
I also found that created tar ball does not include all cpp files.

```
$ tar tvf dist/cupy-2.0.0b1.tar.gz | grep cpp
-rw-rw-r-- ubuntu/ubuntu 5355544 2017-08-29 11:44 cupy-2.0.0b1/cupy/core/core.cpp
```

I verified that `ext_moduels` points cpp files in sources, though.

```
 33 ext_modules = cupy_setup_build.get_ext_modules()
 34 build_ext = cupy_setup_build.custom_build_ext
 35 sdist = cupy_setup_build.sdist_with_cython
 36
 37 for ext in ext_modules:
 38     print(ext.__dict__)
```

```
[omitted]
{'py_limited_api': False, 'name': 'cupy.cuda.memory', 'sources': ['cupy/cuda/memory.cpp'], 'include_dirs': [], 'define_macros': [('_GLIBCXX_USE_CXX11_ABI', '0'), ('_FORCE_INLINES', '1'), ('CUPY_NO_CUDA',
'1')], 'undef_macros': [], 'library_dirs': [], 'libraries': [], 'runtime_library_dirs': [], 'extra_objects': [], 'extra_compile_args': [], 'extra_link_args': [], 'export_symbols': [], 'swig_opts': [], 'de
pends': [], 'language': 'c++', 'optional': None}
[omitted]
```

**Comment by sonots** [comment]
Easy fix is

1. revert as https://github.com/sonots/cupy/commit/b824b45367bd9fdc5e88811ccc28dadfb280fd7d
2. change development procedure from `pip install cython` => `pip install cython fastrlock==0.3`
    * Or, embed all codes of fastrlock into cupy repository as a third party library.

**Comment by sonots** [comment]
I could fix the problem of https://github.com/cupy/cupy/pull/446#issuecomment-325643185 (tar ball does not include cpp files) with a patch https://github.com/sonots/cupy/commit/c7c74a6247ff716243df92689126977f4e8cd86c

**Comment by unnonouno** [comment]
I cannot reproduce SEGV on my machine. Do you know its condition?

**Comment by sonots** [comment]
my env which I could reproduce is as:

```
dev$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.5 LTS"
dev$ python --version
Python 3.6.1
dev$ cython --version
Cython version 0.26
```

no gpu

**Comment by sonots** [comment]
Need to clean everything once with `git clean -fdx`

**Comment by sonots** [comment]
BTW, stacktrace of core is as https://gist.github.com/sonots/947bc653bfba55f2c81b99f7dac24893

**Comment by unnonouno** [comment]
Thank you! With `git clean` i can reproduce the error on my local environment.

**Comment by unnonouno** [comment]
I found that the error occurs here:
https://github.com/cython/cython/blob/master/Cython/Compiler/Scanning.py#L104

This python file is compiled by Cython, and the error is occurred in the compiled file.

**Comment by sonots** [comment]
In my case, it varies.

Python 3.4.6 (as travis):

See https://gist.github.com/sonots/947bc653bfba55f2c81b99f7dac24893#file-gistfile1-txt, it seems error occurs at `Cython/Compiler/Code.c`

See https://gist.github.com/sonots/947bc653bfba55f2c81b99f7dac24893#gistcomment-2190971, SEGV did not occur, but instead Compiler Crash occurs at `from libcpp cimport vector`.

Python 3.6.1:

See https://gist.github.com/sonots/947bc653bfba55f2c81b99f7dac24893#gistcomment-2192673, it seems error occurs at `Cython/Compiler/Scanning.cpython-36m-x86_64-linux-gnu.so` which would similar to the above case.

See https://gist.github.com/sonots/947bc653bfba55f2c81b99f7dac24893#gistcomment-2192675, SEGV did not occur, but instead CompileError occurs at `from cpython.list cimport *`

**Comment by sonots** [comment]
Finally I found the cause. Dynamic `import Cythonize.Build` was doing bad.

https://github.com/sonots/cupy/commit/8be5ae05221f790d399c395b3c4e943cacc274ea

fixed.

**Comment by sonots** [comment]
Could you cheery-pick these commits?

* https://github.com/sonots/cupy/pull/1/commits/c7c74a6247ff716243df92689126977f4e8cd86c
* https://github.com/sonots/cupy/pull/1/commits/8be5ae05221f790d399c395b3c4e943cacc274ea

**Comment by unnonouno** [comment]
There is no SEGV on Travis!!

**Comment by sonots** [comment]
🥇

**Pull request merged**